### ISSUE TITLE
openmpi osc_ucx_component error

### ISSUE THREAD

**Issue by SLgramps** [open-mpi/ompi#12517]
Thank you for taking the time to submit an issue!

## Background information
I have been building/running some simple parallel code (specifically, from https://github.com/modern-fortran, any of the example projects such as tsunami or weather-buoys). When using openmpi v4.1.5 (Fedora 39), everything built/ran as expected. Upon upgrading to Fedora 40, openmpi was also upgraded to v5.0.2 and the code, though running correctly and finishing, issues copious amounts of osc_ucx_component.c:369 errors as shown in the attached screenshot. Downgrading openmpi back to v4.1.5 eliminates the errors.

### What version of Open MPI are you using? (e.g., v4.1.6, v5.0.1, git branch name and hash, etc.)
v4.1.5 (works correctly)
v5.0.2 (issues osc_ucx_component errors)

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)
packages installed from fedora repositories, OpenCoarrays downloaded from opencoarrays.org and built locally
![Screenshot_20240430_142440](https://github.com/open-mpi/ompi/assets/17345594/bb74b184-09ef-4165-a3d3-acd96ddcdf41)

### Please describe the system on which you are running

* Operating system/version: linux kernel 6.8.7 and several predecessors 
* Computer hardware: Intel i7 12700KF, Asus Z-690 motherboard
* Network type:

-----------------------------

## Details of the problem

Please describe, in detail, the problem that you are having, including the behavior you expect to see, the actual behavior that you are seeing, steps to reproduce the problem, etc.  It is most helpful if you can attach a small program that a developer can use to reproduce your problem.

Using v5.0.2, code runs but issues copious osc_ucx_component.c:369 errors as shown in the attached screenshot
Using v4.1.5, code runs without issuing errors
Does not seem to affected by version of linux, gfortran, or OpenCoarrays used

0. install gfortran from fedora repositories, build OpenCoarrays from opencoarrays.org
1. git clone, for instance, https://github.com/modern-fortran/weather-bouys (or tsunami, ch08) and build using included makefiles with either v4.1.5 or v5.02 openmpi
2. run parallel versions of code using cafrun -n 2 (or other number of cores)

(Attached screenshot is from running tsunami/ch08. Code from weather-buoys also exhibits this behavior.)

**Note**: If you include verbatim output (or a code block), please use a [GitHub Markdown](https://help.github.com/articles/creating-and-highlighting-code-blocks/) code block like below:
```shell
shell$ mpirun -n 2 ./hello_world
```

**Comment by janjust** [comment]
as a simple work around just disabling the component will get your around this error `-mca osc ^ucx`

**Comment by SLgramps** [comment]
Yes, this was a local machine. Your suggested workaround eliminates the error. Thanks for looking into this.

**Comment by yizeyi18** [comment]
I'm also facing this issue, and have more information could provide:
This output comes from `ompi/mca/osc/ucx/osc_ucx_component.c:369`, which inferrs `component_set_priority()` implemented in the same file returns a non-success value:

https://github.com/open-mpi/ompi/blob/45258b6cb7129df7c5387445a75f3791cfdc4e74/ompi/mca/osc/ucx/osc_ucx_component.c#L366-L371

And in the `component_set_priority()`, there are 2 `if` that returns fail:

https://github.com/open-mpi/ompi/blob/45258b6cb7129df7c5387445a75f3791cfdc4e74/ompi/mca/osc/ucx/osc_ucx_component.c#L301-L317

Before the second one, `ucp_context_init` would be called, which would use `OSC_UCX_VERBOSE` on ce failed; adjusting environ `OSC_UCX_VERBOSE` and `UCX_LOG_LEVEL` both made no excess output other than line 369, inferring the routine fails on the first `if`.

Before the first `if`, `opal_common_ucx_wpool_allocate()` was called to init `mca_osc_ucx_component.wpool`, which subroutine is implemented in `opal/mca/common/ucx/common_ucx_wpool.c`: 

https://github.com/open-mpi/ompi/blob/45258b6cb7129df7c5387445a75f3791cfdc4e74/opal/mca/common/ucx/common_ucx_wpool.c#L129-L135

in which only `calloc` but no more work was done; this would result as `mca_osc_ucx_component.wpool->ucp_ctx == NULL`, as the first `if` judges. I think there lacks an `opal_common_ucx_wpool_init()` in `ompi/mca/osc/ucx/osc_ucx_component.c:303`.

**Comment by hisacro** [comment]
> as a simple work around just disabling the component will get your around this error `-mca osc ^ucx`

this workwork suggested is giving me errors for mca takes up only single argument, 

```
$ orcaqm -mca "osc ^ucx" phen.inp 
ORCA: too many arguments - 
      Additional mpi parameters have to be given as a single string! 
```

**Comment by devreal** [comment]
@hisacro Try removing the quotes from the parameter, like `--mca osc ^ucx`.

**Comment by hisacro** [comment]
> [@hisacro](https://github.com/hisacro) Try removing the quotes from the parameter, like `--mca osc ^ucx`.

again the same issue, 

```
$ orca --mca osc ^ucx phen2.inp 
ORCA: too many arguments - 
      Additional mpi parameters have to be given as a single string! 
```

**Comment by rhc54** [comment]
Sounds like an issue in this "orca" launcher. Why don't you just use the environmental parameter form of the MCA param and add `OMPI_MCA_osc=^ucx` to your environment?

**Comment by yizeyi18** [comment]
@hisacro If you want strictly using cmdline param, use `$PATH/orca phen2.inp "--mca osc ^ucx"`. MPG ORCA requires MPI params to be at the end of the cmdline, see ORCA 6.1.1 manual sect 2.5.2 at p.42: https://orcaforum.kofo.mpg.de/app.php/dlext/?view=detail&df_id=276

**Comment by schrummy14** [comment]
Hello,
I am also seeing this for 5.0.9. The `--mca osc ^ucx` removed the message for me but reduced the bandwidth between two nodes that I was measuring. This reduction went away if I set the pml to ucx and allowed all devices. I also didn't see the osc error. I also didn't see this issue with 4.1.8.

4.1.8:
```
/opt/JARVICE-MPI/openmpi/bin/mpirun -x OPAL_PREFIX -x LD_LIBRARY_PATH   -n 2 -npernode 1 --hostfile /etc/JARVICE/nodes /usr/local/libexec/osu-micro-benchmarks/mpi/one-sided/osu_get_bw H D 
# OSU MPI_Get Bandwidth Test v7.3
# Window creation: MPI_Win_allocate
# Synchronization: MPI_Win_flush
# Datatype: MPI_CHAR.
# Size      Bandwidth (MB/s)
1                       0.07
2                       0.25
4                       0.35
8                       0.85
16                      2.12
32                      4.13
64                      8.34
128                    16.86
256                    32.25
512                    64.57
1024                  127.03
2048                  251.21
4096                  473.36
8192                  938.48
16384                1236.16
32768                 886.37
65536                1187.02
131072               1189.64
262144               1190.43
524288               1188.57
1048576              1190.08
2097152              1190.72
4194304              1190.87
```

5.0.9 with `--mca osc ^ucx`:
```
/opt/JARVICE-MPI/openmpi/bin/mpirun -x OPAL_PREFIX -x LD_LIBRARY_PATH --mca osc ^ucx  -n 2 -npernode 1 --hostfile /etc/JARVICE/nodes /usr/local/libexec/osu-micro-benchmarks/mpi/one-sided/osu_get_bw H D 
# OSU MPI_Get Bandwidth Test v7.3
# Window creation: MPI_Win_allocate
# Synchronization: MPI_Win_flush
# Datatype: MPI_CHAR.
# Size      Bandwidth (MB/s)
1                       0.09
2                       0.12
4                       0.47
8                       0.97
16                      1.96
32                      3.92
64                      7.78
128                    15.23
256                    29.41
512                    58.50
1024                  116.97
2048                  222.91
4096                  406.10
8192                  671.46
16384                 904.73
32768                1107.21
65536                1191.43
131072                595.20
262144                595.20
524288                595.25
1048576               595.00
2097152               595.22
4194304               594.29
```

5.0.9 forcing UCX allowing any device:
```
/opt/JARVICE-MPI/openmpi/bin/mpirun -x OPAL_PREFIX -x LD_LIBRARY_PATH --mca pml ucx --mca pml_ucx_tls any --mca pml_ucx_devices any  -n 2 -npernode 1 --hostfile /etc/JARVICE/nodes /usr/local/libexec/osu-micro-benchmarks/mpi/one-sided/osu_get_bw H D 
# OSU MPI_Get Bandwidth Test v7.3
# Window creation: MPI_Win_allocate
# Synchronization: MPI_Win_flush
# Datatype: MPI_CHAR.
# Size      Bandwidth (MB/s)
1                       0.10
2                       0.18
4                       0.40
8                       1.05
16                      1.06
32                      4.65
64                      9.37
128                    17.83
256                    35.34
512                    70.64
1024                  140.46
2048                  279.70
4096                  507.60
8192                  764.96
16384                1070.41
32768                1203.70
65536                1186.24
131072               1187.57
262144               1187.95
524288               1188.06
1048576              1188.11
2097152              1187.73
4194304              1188.10
```
